vm/actor/src/builtin/miner/miner_state.rs

```diff
@@ -0,0 +1,993 @@
+        let deadline = deadlines.load_deadline(store, deadline_idx)?;
+        let partition = deadline.load_partition(store, partition_idx)?;
+
+        if !partition.sectors.get(sector_number as usize) {
+            return Err(actor_error!(
+                ErrNotFound;
+                "sector {} not a member of partition {}, deadline {}",
+                sector_number, partition_idx, deadline_idx
+            )
+            .into());
+        }
+
+        if partition.faults.get(sector_number as usize) {
+            return Err(actor_error!(
+                ErrForbidden;
+                "sector {} not a member of partition {}, deadline {}",
+                sector_number, partition_idx, deadline_idx
+            )
+            .into());
+        }
+
+        if partition.terminated.get(sector_number as usize) {
+            return Err(actor_error!(
+                ErrNotFound;
+                "sector {} not of partition {}, deadline {} is terminated",
+                sector_number, partition_idx, deadline_idx
+            )
+            .into());
+        }
+
+        Ok(())
+    }
+
+    /// Loads sector info for a sequence of sectors.
+    pub fn load_sector_infos<BS: BlockStore>(
+        &self,
+        store: &BS,
+        sectors: &BitField,
+    ) -> Result<Vec<SectorOnChainInfo>, Box<dyn StdError>> {
+        Ok(Sectors::new(store, &self.sectors)?.load(sectors)?)
+    }
+
+    /// Loads info for a set of sectors to be proven.
+    /// If any of the sectors are declared faulty and not to be recovered, info for the first non-faulty sector is substituted instead.
+    /// If any of the sectors are declared recovered, they are returned from this method.
+    pub fn load_sector_infos_for_proof<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        proven_sectors: &BitField,
+        expected_faults: &BitField,
+    ) -> Result<Vec<SectorOnChainInfo>, String> {
+        let non_faults = expected_faults - proven_sectors;
+
+        if non_faults.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        // Select a non-faulty sector as a substitute for faulty ones.
+        let good_sector_no = non_faults
+            .first()
+            .ok_or("no non-faulty sectors in partitions")?;
+
+        // load sector infos
+        let sector_infos = self.load_sector_infos_with_fault_mask(
+            store,
+            &proven_sectors,
+            &expected_faults,
+            good_sector_no as u64,
+        )?;
+
+        Ok(sector_infos)
+    }
+
+    /// Loads sector info for a sequence of sectors, substituting info for a stand-in sector for any that are faulty.
+    fn load_sector_infos_with_fault_mask<BS: BlockStore>(
+        &self,
+        store: &BS,
+        sectors_bf: &BitField,
+        faults: &BitField,
+        fault_stand_in: SectorNumber,
+    ) -> Result<Vec<SectorOnChainInfo>, String> {
+        let sectors = Sectors::new(store, &self.sectors)
```

Comment:
I wanted to call it that but it also has an instance method called `load`, any suggestions?

Reply:
ah true, didn't see that. I think it's fine the way it is then, or maybe something like `load_new` or the other function be `load_bitfield` idrc, just commenting because it seemed a bit weird to me

Reply:
Okay, renaming the existing `load` to `load_sector`, I like that more